src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,260 @@
+package ui.components.pickers;
+
+import backend.resource.TurboLabel;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to represent the state of the label picker. In addition to representing a state,
+ * it contains the logic that handles every state transition
+ */
+public class LabelPickerState {
+
+    Set<String> initialLabels;
+    List<String> addedLabels;
+    List<String> removedLabels;
+    List<String> matchedLabels;
+    List<TurboLabel> repoLabels;
+    OptionalInt currentSuggestionIndex;
+
+    @SuppressWarnings("PMD")
+    public LabelPickerState(Set<String> initialLabels, List<TurboLabel> repoLabels, String userInput) {
+        this(initialLabels, new ArrayList<>(), new ArrayList<>(), new ArrayList<>(), repoLabels,
+                OptionalInt.empty());
+        update(userInput);
+    }
+
+    private LabelPickerState(Set<String> initialLabels, List<String> addedLabels, List<String> removedLabels,
+                             List<String> matchedLabels, List<TurboLabel> repoLabels, 
+                             OptionalInt currentSuggestionIndex) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.repoLabels = repoLabels;
+        this.currentSuggestionIndex = currentSuggestionIndex;
+    }
+
+    /**
+     * Updates current state based on given user input
+     * @param userInput 
+     */
+    private void update(String userInput) {
+        List<String> confirmedKeywords = getConfirmedKeywords(userInput);
+        for (String confirmedKeyword : confirmedKeywords) {
+            findAndToggleMatchingLabel(confirmedKeyword);
+        }
+
+        Optional<String> keywordInProgess = getKeywordInProgress(userInput);
+        if (keywordInProgess.isPresent()) {
+            updateMatchedLabels(keywordInProgess.get());
+        }
+    }
+
+    /**
+     * Update current state if there is exactly one matching label based on the 
```

Comment:
Why only if exactly one match? If there are multiple matches, shouldn't we take the first match? (actually, we should take the best match, but we don't have that yet).  Users are supposed to press space as soon as they see the label they want as the 'suggested' label, not when there is only one matched label.